### ISSUE TITLE
Spring Data Repository not taking into account @Field(name) annotation when generating queries for Object Fields

### ISSUE THREAD

**Issue by Mynameisbt** [spring-projects/spring-data-elasticsearch#1752]
I am having an issue when trying to do a search with a Spring Data Elasticsearch 4 repository when using a Object Field Type.  Below is a simple example that is replicating the issue I am having in my application.  With Spring Data Elasticsearch 3 the query generated by the repository would use the `@Field(name)` value but it no longer does with Spring Data Elasticsearch 4.  

The query that it is generating for the Authors.FirstName field repository method according to the logs is:

Request body: {"from":0,"size":10000,"query":{"bool":{"must":[{"query_string":{"query":"John","fields":["authors.firstName^1.0"],"type":"best_fields","default_operator":"and","max_determinized_states":10000,"enable_position_increments":true,"fuzziness":"AUTO","fuzzy_prefix_length":0,"fuzzy_max_expansions":50,"phrase_slop":0,"escape":false,"auto_generate_synonyms_phrase_query":true,"fuzzy_transpositions":true,"boost":1.0}}],"adjust_pure_negative":true,"boost":1.0}},"version":true}

I would have expected the fields query to be:

["Authors.FirstName^1.0"]  

based on the `@Field` annotation.  That is what Spring Data Elasticsearch 3 would generate

Is there something I am missing in my configuration for the query to use the `@Field(name)` value in Spring Data Elasticsearch 4?

I have a github test setup that reproduces the problem.  I just run a Elasticsearch docker container prior to executing the test.

https://github.com/Mynameisbt/nestedobjectquerytest/blob/master/src/test/java/com/mynameisbt/nestedobjectquerytest/NestedobjectquerytestApplicationTests.java

If I remove the custom names for authors and first name then the documents get returned.

```
@Document(indexName = "blog", type = "article")
public class Article {
    @Id
    private String id;
    @MultiField(mainField = @Field(type = Text, fielddata = true), otherFields = {@InnerField(suffix = "verbatim", type = Keyword)})
    private String title;

    @Field(type = FieldType.Object, name = "Authors")
    private List<Author> authors;

    @Field(type = Keyword)
    private String[] tags;

    public Article() {
    }

    public Article(String title) {
        this.title = title;
    }
}
```
Subclass
```
public class Author {
    @Field("FirstName")
    private String firstName;
    @Field("LastName")
    private String lastName;

```

Repository
```
public interface ArticleRepository extends ElasticsearchRepository<Article, String> {

    Page<Article> findByAuthorsFirstName(String name, Pageable pageable);
}
```

**Comment by sothawo** [comment]
The problem is that the `@Field` annotations of the properties of `Author` do not have a `type` attribute set, meaning the filed type for Elasticsearch is autodetected and therefore no mapping is written - even if a `name`was specified. You will need to change that to:

```java
public class Author {
    @Field(value = "FirstName", type = FieldType.Text)
    private String firstName;
    @Field(value = "LastName", type = FieldType.Text)
    private String lastName;
}
```

I will add a description of this behaviour to the documentation.

But even with this change, the test will fail, because currently the `CriteriaQuery` - and the repository implementation creates one under the hood - does not support queries on nested objects [see my answer on this SO question](https://stackoverflow.com/a/66698200/4393565) where this came up as well.

I will later create a bug issue for this, as this should definitely be possible.

**Comment by sothawo** [comment]
Forget the first part of my comment with the field type, that's not needed here for the mapping. The data is stored correctly with the capitalized field names. The problem is only with the nested query.

**Comment by sothawo** [comment]
Not having the `type` defined in the `@Field` annotations is no problem. The field won't be written to the index mapping on index creation,  but later will be autodetected as a text field with the desired name (_FirstName, LastName_).

But the problem really is the nested query. I have implemented a fix with ~#1753~, this will be available from the next version on - which will be Spring Data Elasticsearch 4.2.0 GA. 

Your demo application successfully runs with this version, what must be changed is:
```java
@Field(type = FieldType.Nested, name = "Authors")
private List<Author> authors;
```
The type must be changed to `FieldType.Nested` as we need a nested query in this case.

**Comment by Mynameisbt** [comment]
Shouldn't it also pass even if the FieldType  is FieldType.Object?

 I know in the Author example it makes sense to use FieldType.Nested and a nested query for names to maintain the independence of each object vs FieldType.Object where the objects are flatted.  https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html

But in my example even with a FieldType.Object query where the object arrays are flattened I would expect it to return the same 4 documents since the query is just on the FirstName field which is on all 4 documents.

**Comment by sothawo** [comment]
I added another fix ~#1761~ so now a nested query is only used when the `FieldType.Nested` is defined, otherwise a normal query is used.
Your demo code runs now with no change - at least no change relevant to the problem, you will need to remove the `type` attribute from the `@Document` annotation.

**Comment by Mynameisbt** [comment]
Thanks for making the fixes.  Will these fixes also be added to 4.0.9 and 4.1.8 releases or will I need to update to version 4.2?

**Comment by sothawo** [comment]
I will check this evening if this can be backported
